tooling/greybox_fuzzer/src/lib.rs

```diff
@@ -271,10 +274,13 @@ pub struct FuzzedExecutorExecutionConfiguration {
 }
 
 pub struct FuzzedExecutorFailureConfiguration {
-    /// Fail on specific asserts
+    /// Fail on specific asserts (when you use `fuzz(only_fail_with = "...")`)
     pub fail_on_specific_asserts: bool,
 
-    /// Failure reason
+    /// We expect the fuzzed program to always fail (when you use `fuzz(should_fail)` or `fuzz(should_fail_with = "...")` )
+    pub should_always_fail: bool,
+
+    /// Failure reason (when you use `fuzz(should_fail_with = "...")` or `fuzz(only_fail_with = "...")`)
     pub failure_reason: Option<String>,
 }
```

Comment:
I wonder if it wouldn't be better to have this configuration be an enum, similar to `FuzzingScope`, but maybe with four variants: `None`, `OnlyFailWith`, `ShouldFail` and `ShouldFailWith`. Then in `single_fuzz_...` we could pattern match each of these and handle them separately. Mainly because right now we get the failure reason and then assume one of the other booleans is true, adding `unreachable!` otherwise. I think with pattern-matching we wouldn't have an unreachable state and it might also lead to clearer code (right now we are checking these booleans many times).

Reply:
Changed